### DISCUSSION TITLE
Serverless authenticated patch/post/put requests failing because of CORS

### DISCUSSION THREAD

**Discussion by GMaiolo** [vercel/vercel#4064] [General]
Hi, I'm not able to correctly handle CORS issues when doing either `PATCH`/`POST`/`PUT` requests from the browser.

This is my now.json headers section

![07e14526-793f-4dbc-9774-be2755a5825e-code](https://user-images.githubusercontent.com/9776436/79185323-f5ee2400-7dec-11ea-9fc1-e1e3d9ad76f0.png)

I tried using npm cors package without success, also tried to add `routes` in the now.json but without success and also tried handling `OPTIONS` request manually doing

![6f0904e5-5ee7-455b-8bde-e4f40f818309-code](https://user-images.githubusercontent.com/9776436/79185352-04d4d680-7ded-11ea-854b-1359677f5783.png)

Finally, this is the error that I get

```
Access to XMLHttpRequest at 'https://api.example.org/api/users' from origin 'https://example.org' has been blocked by CORS policy: 
Response to preflight request doesn't pass access control check: It does not have HTTP ok status.
```

I don't really know what do try anymore, anyone got hints?

**Comment by coetry** [comment]
Hi @GMaiolo I was able to bypass any CORS issues without needing an external npm package and have an API deployed to [api.hai.af](https://api.hai.af) and also a page that queries the API from another domain deployed at [hai.af/greeting](https://hai.af/greeting)

I have my logs public for both deployments so you can take a closer look:

[API now.json](https://zeit.co/coetry/api/qfsvwvatr/source)
[Querying Page](https://zeit.co/coetry/haiaf/b4sjrn8mi/source?f=src%2Fpages%2Fgreeting%2Findex.js&h=219%3B51)

**Reply by GMaiolo** [reply]
Hey @coetry, my issue is not with `GET` requests but with `PATCH`/`POST`/`PUT` when using an `Authorization` header with a bearer token, this basically triggers an `OPTIONS` request to the server and it seems that it's not handled automatically by Zeit.

Your example does a `GET` request and that same code works for me without any issues.

I found a workaround right now using `micro-cors` library, I wrote a bit about it in my own answer in [this Stackoverflow question](https://stackoverflow.com/questions/61217119/zeit-now-serverless-authenticated-requests-failing-because-of-cors/61242176).

Thanks anyways, if you have more info I'd be glad to hear it as I'm puzzled about why Zeit doesn't handle this out of the box (or if it was me who was missing something).

**Comment by GMaiolo** [comment]
[Link to Stackoverflow question](https://stackoverflow.com/questions/61217119/zeit-now-serverless-authenticated-requests-failing-because-of-cors/61242176) with workaround for this issue.

Even though I was able to workaround this It'd be nice to understand why this happened and how to avoid manually fixing it as it's safe to assume that the headers in `now.json` should suffice.

**Comment by zackkrida** [comment]
Wow, EXACT same issue, wildly annoying. Thanks for posting the workaround
I've never seen the particular `It does not have HTTP ok status.` with a preflight request before

**Comment by khalby786** [comment]
Unfortunately, the workaround isn't also working for me. I'm in the same situation as the OP, with a Bearer token in the Authorization header. Tried all the methods to enable CORS using the methods suggested by Vercel themselves. Would appreciate it if someone could tell me if there's another possible way to fix this.

**Reply by GMaiolo** [reply]
Are you using any expressjs middleware or any external middlewares?

**Reply by khalby786** [reply]
Update after many months 🙂:

It turned out that the way I was sending response from my REST API was incorrect, which resulted in errors causing the request to return a CORS error rather than a Vercel error message. 

To be specific, I used `res.sendStatus` instead of `res.status()`.

**Reply by GMaiolo** [reply]
Glad it worked at least!

**Comment by krehwell** [comment]
hey there author, are this things also applied to vercel? I saw it is made for zeit. 

I have the same problem where my website keep blocking me from requesting my rest-api. is there any other workaround you found?

**Reply by GMaiolo** [reply]
Hey, as far as I know it does apply to Vercel.
What's your scenario?

**Reply by krehwell** [reply]
may I know how do you apply it in the rest api? I am using express for my server.

when my site trying to make a request using "PUT" the errors says that Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://something.com/user/remove-cookies. (Reason: CORS preflight response did not succeed).

I have made a vercel.json to allow my website to be access the rest-api but still dont know how to solve this error. I cant find any similar problem on internet. you problem looks the most similar one.

my vercel.json is the same with you with additional that other headers properties. I tried to use the micro-cors but I could be implement it wrong which makes it does now work. I basically just copy the the code and export it then use it in the index.js as belo

```
// corz.js
import micro from "micro-cors";

function MyApi(req, res) {
  if (req.method === "OPTIONS") {
    return res.status(200).end();
  }
  // handling other requests normally after this
}

const cors = micro();

export default cors(MyApi);
```

```
// index.js
const express = require('express');
const app = express();
const corz = require('./corz.js');

app.use(corz());
```

is my implementation correct tho? how to implement the micro-cors? .


**edit**: ahh my bad I think, I thought micro-cors is just a name made to allow cors in vercel but turns out that micro is just another node tools to handle http. do you have any workaround to allow it using express?

**Reply by GMaiolo** [reply]
Can you try doing `app.use(corz);` instead of `corz()`? Just a wild wild attempt.
I don't quite remember, but does Vercel still support a live express server? I thought it didn't anymore.

Also — are you using some kind of authentication middleware or any other middlewares?

**Reply by krehwell** [reply]
just tried it but still not working. I can do a GET request but that what made me believe there is actually a solution for this.
> Also — are you using some kind of authentication middleware or any other middlewares  

there will be if I can login first to my dashboard. but for the request I told you there is no auth middleware for it. its just a simple login request where the server send back the cookie to the site
